midend/flattenUnions.cpp

```diff
@@ -0,0 +1,359 @@
+                                           const IR::Member *m, const IR::Type_HeaderUnion *hu,
+                                           cstring exclude, bool setValidforCurrMem) {
+    auto code_block = new IR::IndexedVector<IR::StatOrDecl>;
+    if (setValidforCurrMem) {
+        code_block->push_back(processValidityForStr(s, m, exclude, IR::Type_Header::setValid));
+    }
+
+    code_block->push_back(s);
+
+    for (auto sfu : hu->fields) {
+        if (exclude != sfu->name.name) {  // setInvalid for rest of the fields
+            code_block->push_back(processValidityForStr(s, m, sfu->name,
+                                                        IR::Type_Header::setInvalid));
+        }
+    }
+    return new IR::BlockStatement(*code_block);
+}
+
+const IR::Node * HandleValidityHeaderUnion::expandIsValid(
+                 const IR::Statement *a, const IR::MethodCallExpression *mce) {
+    auto mi = P4::MethodInstance::resolve(mce, refMap, typeMap);
+    if (auto bim = mi->to<P4::BuiltInMethod>()) {
+        if (bim->name == "isValid") {   // hdr.u.isValid() or u.isValid
+            if (auto huType = bim->appliedTo->type->to<IR::Type_HeaderUnion>()) {  // u or hdr.u
+                cstring tmp = refMap->newName("tmp");
+                IR::PathExpression *tmpVar = new IR::PathExpression(IR::ID(tmp));
+                toInsert.push_back(new IR::Declaration_Variable(
+                                   IR::ID(tmp), IR::Type_Bits::get(32)));
+                auto code_block = new IR::IndexedVector<IR::StatOrDecl>;
+                code_block->push_back(new IR::AssignmentStatement(a->srcInfo,
+                                      tmpVar, new IR::Constant(IR::Type_Bits::get(32), 0)));
+                for (auto sfu : huType->fields) {
+                    auto method = new IR::Member(a->srcInfo,
+                                                  new IR::Member(bim->appliedTo, sfu->name),
+                                                  IR::ID(IR::Type_Header::isValid));
+                    auto mc = new IR::MethodCallExpression(a->srcInfo, method,
+                                                           new IR::Vector<IR::Argument>());
+                    typeMap->setType(method, new IR::Type_Method(IR::Type_Void::get(),
+                                     new IR::ParameterList(), IR::Type_Header::isValid));
+                    auto addOp = new IR::Add(a->srcInfo, tmpVar,
+                                             new IR::Constant(IR::Type_Bits::get(32), 1));
+                    auto assn = new IR::AssignmentStatement(a->srcInfo, tmpVar, addOp);
+                    code_block->push_back(new IR::IfStatement(a->srcInfo, mc, assn, nullptr));
+                }
+                auto cond = new IR::Equ(a->srcInfo, tmpVar,
+                                        new IR::Constant(IR::Type_Bits::get(32), 1));
+                if (auto  assn = a->to<IR::AssignmentStatement>())
```

Comment:
can you return a value and let the postorder caller do this part?

Reply:
Done.